### ISSUE TITLE
Dramatic slowdown for 2D gamma since version 0.4.3

### ISSUE THREAD

**Issue by SimonBiggs** [SimonBiggs/npgamma#11]
Upon using npgamma today I noticed that the most recent version is dramatically slower for 2D gamma calculation than version 0.4.3. https://pypi.python.org/pypi/npgamma/0.4.3.

Need to get to the bottom of the slow down (either the multithreading or the RAM control). And also put a test in place to check that updates don't don't dramatically slow down calculation.

**Comment by SimonBiggs** [comment]
Comparing [0.4.3](https://pypi.python.org/packages/33/19/f967878aeba665597c175367b22d96ab01b27ef5b141daa7c93b4430d555/npgamma-0.4.3.zip#md5=bc182680c39fb0e5cb8232114cdfcf3d) and [0.5.2](https://pypi.python.org/packages/ac/a7/5025de5f2edd9b6dea94d95f004eb23b17ab50e0afc44829a1a399ba2ac4/npgamma-0.5.2.tar.gz#md5=520b857accc100cf62666cc1d5d284f8). The issue is emphasised when the eventual gamma value requires finding points a large distance from the evaluation point. In the case below the run time of 0.5.2 is 18x slower than 0.4.3.

The code for the respective libraries are on pypi (https://pypi.python.org/simple/npgamma/) 

The code used for the comparison is the following:

```
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline

from npgamma import calc_gamma


grid = 0.01
scale_factor = 1.3
noise = 0.01


x = np.arange(-1.8, 1.8 + 0.001, grid)
y = np.arange(-1.5, 1.5 + 0.001, grid)
coords = (y, x)

xx, yy = np.meshgrid(x, y)

dose_ref = np.exp(-(xx**10 + yy**10))
dose_eval = dose_ref * scale_factor

## ipython cell magic
%%prun -s cumulative

gamma_no_noise = calc_gamma(
    coords, dose_ref,
    coords, dose_eval,
    0.03, 0.03, lower_dose_cutoff=0.2)
```


## Profile for 0.4.3

```
         74493 function calls in 4.748 seconds

   Ordered by: cumulative time

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.000    0.000    4.748    4.748 {built-in method builtins.exec}
        1    0.000    0.000    4.748    4.748 <string>:3(<module>)
        1    0.001    0.001    4.748    4.748 main.py:19(calc_gamma)
        1    0.026    0.026    4.746    4.746 main.py:267(calc_gamma_2d)
       20    0.142    0.007    4.713    0.236 main.py:163(find_min_dose_difference_at_distance_2d)
       20    0.045    0.002    4.278    0.214 main.py:176(<listcomp>)
     1204    0.212    0.000    4.233    0.004 interpolate.py:2325(__call__)
     1204    1.801    0.001    2.178    0.002 interpolate.py:2370(_evaluate_linear)
     1204    0.845    0.001    1.798    0.001 interpolate.py:2391(_find_indices)
     2408    0.002    0.000    0.946    0.000 fromnumeric.py:1036(searchsorted)
     2408    0.944    0.000    0.944    0.000 {method 'searchsorted' of 'numpy.ndarray' objects}
     9632    0.326    0.000    0.326    0.000 {built-in method numpy.core.multiarray.where}
       40    0.186    0.005    0.186    0.005 {built-in method numpy.core.multiarray.concatenate}
     4958    0.078    0.000    0.078    0.000 {method 'reduce' of 'numpy.ufunc' objects}
    10908    0.055    0.000    0.055    0.000 {built-in method numpy.core.multiarray.array}
       80    0.000    0.000    0.054    0.001 fromnumeric.py:2300(amin)
       80    0.000    0.000    0.054    0.001 _methods.py:28(_amin)
     1204    0.046    0.000    0.046    0.000 interpolate.py:2376(<listcomp>)
     4818    0.007    0.000    0.039    0.000 fromnumeric.py:1987(all)
     4818    0.004    0.000    0.028    0.000 {method 'all' of 'numpy.ndarray' objects}
     4818    0.002    0.000    0.025    0.000 _methods.py:40(_all)
     1204    0.006    0.000    0.006    0.000 {built-in method numpy.core.multiarray.zeros}
     4821    0.004    0.000    0.005    0.000 numeric.py:414(asarray)
     6066    0.004    0.000    0.005    0.000 numeric.py:484(asanyarray)
     1204    0.003    0.000    0.004    0.000 {built-in method scipy.interpolate.interpnd._ndim_coords_from_arrays}
     2410    0.003    0.000    0.003    0.000 {method 'reshape' of 'numpy.ndarray' objects}
       20    0.000    0.000    0.002    0.000 fromnumeric.py:1743(sum)
       20    0.000    0.000    0.002    0.000 _methods.py:31(_sum)
       20    0.000    0.000    0.001    0.000 shape_base.py:180(vstack)
       20    0.001    0.000    0.001    0.000 main.py:109(coords_to_check_2d)
        1    0.000    0.000    0.001    0.001 function_base.py:3979(meshgrid)
     4856    0.001    0.000    0.001    0.000 {method 'append' of 'list' objects}
        1    0.001    0.001    0.001    0.001 function_base.py:4116(<listcomp>)
       20    0.000    0.000    0.001    0.000 function_base.py:9(linspace)
     3747    0.001    0.000    0.001    0.000 {built-in method builtins.len}
        3    0.001    0.000    0.001    0.000 {built-in method numpy.core.multiarray.copyto}
        2    0.000    0.000    0.000    0.000 numeric.py:194(ones_like)
       20    0.000    0.000    0.000    0.000 shape_base.py:230(<listcomp>)
       40    0.000    0.000    0.000    0.000 shape_base.py:61(atleast_2d)
       40    0.000    0.000    0.000    0.000 fromnumeric.py:2200(amax)
       20    0.000    0.000    0.000    0.000 {built-in method numpy.core.multiarray.arange}
       21    0.000    0.000    0.000    0.000 {method 'astype' of 'numpy.ndarray' objects}
        1    0.000    0.000    0.000    0.000 numeric.py:148(ones)
       40    0.000    0.000    0.000    0.000 _methods.py:25(_amax)
        1    0.000    0.000    0.000    0.000 interpolate.py:2284(__init__)
       20    0.000    0.000    0.000    0.000 {built-in method numpy.core.multiarray.result_type}
       20    0.000    0.000    0.000    0.000 {built-in method builtins.isinstance}
        2    0.000    0.000    0.000    0.000 function_base.py:1518(diff)
        2    0.000    0.000    0.000    0.000 {built-in method numpy.core.multiarray.empty_like}
        1    0.000    0.000    0.000    0.000 function_base.py:4096(<listcomp>)
        1    0.000    0.000    0.000    0.000 numerictypes.py:728(issubdtype)
        1    0.000    0.000    0.000    0.000 {built-in method numpy.core.multiarray.empty}
        1    0.000    0.000    0.000    0.000 main.py:47(<listcomp>)
        4    0.000    0.000    0.000    0.000 fromnumeric.py:1574(shape)
        1    0.000    0.000    0.000    0.000 {built-in method numpy.core.multiarray.can_cast}
        1    0.000    0.000    0.000    0.000 interpolate.py:2322(<listcomp>)
        1    0.000    0.000    0.000    0.000 numerictypes.py:660(issubclass_)
        1    0.000    0.000    0.000    0.000 function_base.py:4099(<listcomp>)
        4    0.000    0.000    0.000    0.000 {built-in method builtins.hasattr}
        1    0.000    0.000    0.000    0.000 main.py:53(<listcomp>)
        2    0.000    0.000    0.000    0.000 {built-in method builtins.issubclass}
        3    0.000    0.000    0.000    0.000 {method 'pop' of 'dict' objects}
        1    0.000    0.000    0.000    0.000 {method 'disable' of '_lsprof.Profiler' objects}
```


## Profile for 0.5.2

```
         1993787 function calls in 91.932 seconds

   Ordered by: cumulative time

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.000    0.000   91.932   91.932 {built-in method builtins.exec}
        1    0.000    0.000   91.932   91.932 <string>:3(<module>)
        1    0.000    0.000   91.932   91.932 main.py:28(calc_gamma)
        1    0.000    0.000   91.932   91.932 main.py:141(__init__)
        1    0.072    0.072   91.931   91.931 main.py:195(calculate_gamma)
      102    0.131    0.001   91.832    0.900 main.py:239(calculate_min_dose_difference_by_slice)
      102    2.719    0.027   91.455    0.897 main.py:270(calculate_min_dose_difference)
      102    2.532    0.025   81.107    0.795 main.py:282(<listcomp>)
    32518    4.113    0.000   78.575    0.002 interpolate.py:2325(__call__)
    32518   32.844    0.001   40.517    0.001 interpolate.py:2370(_evaluate_linear)
    32518   15.530    0.000   32.788    0.001 interpolate.py:2391(_find_indices)
    65036    0.040    0.000   17.079    0.000 fromnumeric.py:1036(searchsorted)
    65036   17.039    0.000   17.039    0.000 {method 'searchsorted' of 'numpy.ndarray' objects}
   260246    6.683    0.000    6.683    0.000 {built-in method numpy.core.multiarray.where}
   293595    3.449    0.000    3.449    0.000 {built-in method numpy.core.multiarray.array}
      204    3.299    0.016    3.299    0.016 {built-in method numpy.core.multiarray.concatenate}
   130992    1.496    0.000    1.496    0.000 {method 'reduce' of 'numpy.ufunc' objects}
   130278    0.184    0.000    0.968    0.000 fromnumeric.py:1987(all)
      408    0.002    0.000    0.948    0.002 fromnumeric.py:2300(amin)
      408    0.000    0.000    0.946    0.002 _methods.py:28(_amin)
    32518    0.870    0.000    0.870    0.000 interpolate.py:2376(<listcomp>)
   130278    0.083    0.000    0.672    0.000 {method 'all' of 'numpy.ndarray' objects}
   130278    0.051    0.000    0.588    0.000 _methods.py:40(_all)
    32620    0.156    0.000    0.156    0.000 {built-in method numpy.core.multiarray.zeros}
   163106    0.119    0.000    0.150    0.000 numeric.py:484(asanyarray)
   130077    0.121    0.000    0.146    0.000 numeric.py:414(asarray)
      102    0.001    0.000    0.135    0.001 fromnumeric.py:717(sort)
      102    0.129    0.001    0.129    0.001 {method 'sort' of 'numpy.ndarray' objects}
    32518    0.066    0.000    0.103    0.000 {built-in method scipy.interpolate.interpnd._ndim_coords_from_arrays}
    65038    0.077    0.000    0.077    0.000 {method 'reshape' of 'numpy.ndarray' objects}
      102    0.042    0.000    0.043    0.000 {method 'shuffle' of 'mtrand.RandomState' objects}
   130582    0.023    0.000    0.023    0.000 {method 'append' of 'list' objects}
    98589    0.020    0.000    0.020    0.000 {built-in method builtins.len}
      102    0.015    0.000    0.015    0.000 main.py:258(<listcomp>)
      102    0.000    0.000    0.013    0.000 fromnumeric.py:1743(sum)
      102    0.000    0.000    0.012    0.000 _methods.py:31(_sum)
      102    0.004    0.000    0.006    0.000 main.py:290(calculate_coordinates_kernel)
      102    0.000    0.000    0.006    0.000 shape_base.py:180(vstack)
      102    0.004    0.000    0.004    0.000 {method 'copy' of 'numpy.ndarray' objects}
      204    0.003    0.000    0.003    0.000 {built-in method numpy.core.multiarray.arange}
      102    0.002    0.000    0.003    0.000 shape_base.py:378(array_split)
      104    0.000    0.000    0.003    0.000 numeric.py:194(ones_like)
      207    0.003    0.000    0.003    0.000 {built-in method numpy.core.multiarray.copyto}
      102    0.001    0.000    0.002    0.000 function_base.py:9(linspace)
      102    0.000    0.000    0.002    0.000 shape_base.py:230(<listcomp>)
      204    0.001    0.000    0.001    0.000 shape_base.py:61(atleast_2d)
      102    0.000    0.000    0.001    0.000 numeric.py:86(zeros_like)
      205    0.001    0.000    0.001    0.000 {method 'astype' of 'numpy.ndarray' objects}
      204    0.000    0.000    0.001    0.000 fromnumeric.py:2200(amax)
      204    0.000    0.000    0.001    0.000 _methods.py:25(_amax)
        1    0.000    0.000    0.001    0.001 function_base.py:3979(meshgrid)
      206    0.000    0.000    0.000    0.000 {built-in method numpy.core.multiarray.empty_like}
      102    0.000    0.000    0.000    0.000 {method 'cumsum' of 'numpy.ndarray' objects}
        1    0.000    0.000    0.000    0.000 function_base.py:4116(<listcomp>)
      202    0.000    0.000    0.000    0.000 _internal.py:227(__init__)
      204    0.000    0.000    0.000    0.000 fromnumeric.py:456(swapaxes)
      104    0.000    0.000    0.000    0.000 {built-in method builtins.isinstance}
      102    0.000    0.000    0.000    0.000 {built-in method numpy.core.multiarray.result_type}
      204    0.000    0.000    0.000    0.000 {method 'swapaxes' of 'numpy.ndarray' objects}
        1    0.000    0.000    0.000    0.000 interpolate.py:2284(__init__)
      102    0.000    0.000    0.000    0.000 {built-in method builtins.divmod}
        1    0.000    0.000    0.000    0.000 numeric.py:148(ones)
      202    0.000    0.000    0.000    0.000 _internal.py:252(get_data)
        2    0.000    0.000    0.000    0.000 function_base.py:1518(diff)
        1    0.000    0.000    0.000    0.000 function_base.py:4096(<listcomp>)
        1    0.000    0.000    0.000    0.000 numerictypes.py:728(issubdtype)
        1    0.000    0.000    0.000    0.000 {built-in method numpy.core.multiarray.can_cast}
        1    0.000    0.000    0.000    0.000 numerictypes.py:660(issubclass_)
        1    0.000    0.000    0.000    0.000 interpolate.py:2322(<listcomp>)
        4    0.000    0.000    0.000    0.000 fromnumeric.py:1574(shape)
        1    0.000    0.000    0.000    0.000 main.py:82(<listcomp>)
        1    0.000    0.000    0.000    0.000 {built-in method numpy.core.multiarray.empty}
        4    0.000    0.000    0.000    0.000 {built-in method builtins.hasattr}
        3    0.000    0.000    0.000    0.000 {method 'pop' of 'dict' objects}
        1    0.000    0.000    0.000    0.000 {method 'disable' of '_lsprof.Profiler' objects}
        1    0.000    0.000    0.000    0.000 function_base.py:4099(<listcomp>)
        1    0.000    0.000    0.000    0.000 main.py:88(<listcomp>)
        2    0.000    0.000    0.000    0.000 {built-in method builtins.issubclass}
```

**Comment by SimonBiggs** [comment]
The issue was simply that from upgrading from 0.4.3 to 0.5.2 I changed the default setting of "maximum test distance" from 2 times distance threshold, to instead being np.inf. I decided that a maximum test distance should not be set unless the user explicitly set it themselves. As a result there is a significant increase in calculation times using the corresponding default settings. Especially when there is large gamma values. This can be rectified if one wishes by setting `maximum_test_distance=distance_threshold * 2`.

Apart from that the speed difference is only small.
